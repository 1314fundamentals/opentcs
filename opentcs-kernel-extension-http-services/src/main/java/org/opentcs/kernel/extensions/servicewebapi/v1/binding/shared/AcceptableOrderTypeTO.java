// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import java.util.Objects;

public class AcceptableOrderTypeTO
    implements
      Serializable {

  private final String name;
  private final int priority;

  @JsonCreator
  public AcceptableOrderTypeTO(
      @JsonProperty(value = "name", required = true)
      String name,
      @JsonProperty(value = "priority", required = true)
      int priority
  ) {
    this.name = requireNonNull(name, "name");
    this.priority = priority;
  }

  public String getName() {
    return name;
  }

  public int getPriority() {
    return priority;
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (!(o instanceof AcceptableOrderTypeTO other)) {
      return false;
    }

    return Objects.equals(name, other.getName()) && priority == other.getPriority();
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, priority);
  }

  @Override
  public String toString() {
    return "AcceptableOrderTypeTO{"
        + "name=" + name
        + ", priority=" + priority
        + '}';
  }

}
