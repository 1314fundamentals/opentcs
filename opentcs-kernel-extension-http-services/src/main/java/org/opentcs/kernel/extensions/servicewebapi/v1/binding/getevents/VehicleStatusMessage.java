// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import org.opentcs.data.model.Vehicle;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.AcceptableOrderTypeTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.BoundingBoxTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.CoupleTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * A status message containing information about a vehicle.
 */
public class VehicleStatusMessage
    extends
      StatusMessage {

  private String vehicleName = "";

  private List<Property> properties = new ArrayList<>();

  private String transportOrderName = "";

  private BoundingBoxTO boundingBox = new BoundingBoxTO(
      1000,
      1000,
      1000,
      new CoupleTO(0, 0)
  );

  private int energyLevelGood;

  private int energyLevelCritical;

  private int energyLevelSufficientlyRecharged;

  private int energyLevelFullyRecharged;

  private int energyLevel;

  private Vehicle.IntegrationLevel integrationLevel = Vehicle.IntegrationLevel.TO_BE_RESPECTED;

  private String position;

  private PrecisePosition precisePosition;

  private double orientationAngle;

  private boolean paused;

  private Vehicle.State state;

  private Instant stateTimestamp = Instant.EPOCH;

  private Vehicle.ProcState procState;

  private Instant procStateTimestamp = Instant.EPOCH;

  private List<List<String>> allocatedResources = new ArrayList<>();

  private List<List<String>> claimedResources = new ArrayList<>();

  private List<AcceptableOrderTypeTO> acceptableOrderTypes = new ArrayList<>();

  private String envelopeKey;

  /**
   * Creates a new instance.
   */
  public VehicleStatusMessage() {
  }

  @Override
  public VehicleStatusMessage setSequenceNumber(long sequenceNumber) {
    return (VehicleStatusMessage) super.setSequenceNumber(sequenceNumber);
  }

  @Override
  public VehicleStatusMessage setCreationTimeStamp(Instant creationTimeStamp) {
    return (VehicleStatusMessage) super.setCreationTimeStamp(creationTimeStamp);
  }

  public String getVehicleName() {
    return vehicleName;
  }

  public VehicleStatusMessage setVehicleName(String vehicleName) {
    this.vehicleName = vehicleName;
    return this;
  }

  public List<Property> getProperties() {
    return properties;
  }

  public VehicleStatusMessage setProperties(List<Property> properties) {
    this.properties = properties;
    return this;
  }

  public String getTransportOrderName() {
    return transportOrderName;
  }

  public VehicleStatusMessage setTransportOrderName(String transportOrderName) {
    this.transportOrderName = transportOrderName;
    return this;
  }

  @Nonnull
  public BoundingBoxTO getBoundingBox() {
    return boundingBox;
  }

  public VehicleStatusMessage setBoundingBox(
      @Nonnull
      BoundingBoxTO boundingBox
  ) {
    this.boundingBox = requireNonNull(boundingBox, "boundingBox");
    return this;
  }

  public int getEnergyLevelGood() {
    return energyLevelGood;
  }

  public VehicleStatusMessage setEnergyLevelGood(int energyLevelGood) {
    this.energyLevelGood = energyLevelGood;
    return this;
  }

  public int getEnergyLevelCritical() {
    return energyLevelCritical;
  }

  public VehicleStatusMessage setEnergyLevelCritical(int energyLevelCritical) {
    this.energyLevelCritical = energyLevelCritical;
    return this;
  }

  public int getEnergyLevelSufficientlyRecharged() {
    return energyLevelSufficientlyRecharged;
  }

  public VehicleStatusMessage setEnergyLevelSufficientlyRecharged(
      int energyLevelSufficientlyRecharged
  ) {
    this.energyLevelSufficientlyRecharged = energyLevelSufficientlyRecharged;
    return this;
  }

  public int getEnergyLevelFullyRecharged() {
    return energyLevelFullyRecharged;
  }

  public VehicleStatusMessage setEnergyLevelFullyRecharged(int energyLevelFullyRecharged) {
    this.energyLevelFullyRecharged = energyLevelFullyRecharged;
    return this;
  }

  public int getEnergyLevel() {
    return energyLevel;
  }

  public VehicleStatusMessage setEnergyLevel(int energyLevel) {
    this.energyLevel = energyLevel;
    return this;
  }

  public Vehicle.IntegrationLevel getIntegrationLevel() {
    return integrationLevel;
  }

  public VehicleStatusMessage setIntegrationLevel(Vehicle.IntegrationLevel integrationLevel) {
    this.integrationLevel = requireNonNull(integrationLevel, "integrationLevel");
    return this;
  }

  public String getPosition() {
    return position;
  }

  public VehicleStatusMessage setPosition(String position) {
    this.position = position;
    return this;
  }

  public PrecisePosition getPrecisePosition() {
    return precisePosition;
  }

  public VehicleStatusMessage setPrecisePosition(PrecisePosition precisePosition) {
    this.precisePosition = precisePosition;
    return this;
  }

  public double getOrientationAngle() {
    return orientationAngle;
  }

  public VehicleStatusMessage setOrientationAngle(double orientationAngle) {
    this.orientationAngle = orientationAngle;
    return this;
  }

  public boolean isPaused() {
    return paused;
  }

  public VehicleStatusMessage setPaused(boolean paused) {
    this.paused = paused;
    return this;
  }

  public Vehicle.State getState() {
    return state;
  }

  public VehicleStatusMessage setState(Vehicle.State state) {
    this.state = state;
    return this;
  }

  public Instant getStateTimestamp() {
    return stateTimestamp;
  }

  public VehicleStatusMessage setStateTimestamp(Instant stateTimestamp) {
    this.stateTimestamp = requireNonNull(stateTimestamp, "stateTimestamp");
    return this;
  }

  public Vehicle.ProcState getProcState() {
    return procState;
  }

  public VehicleStatusMessage setProcState(Vehicle.ProcState procState) {
    this.procState = procState;
    return this;
  }

  public Instant getProcStateTimestamp() {
    return procStateTimestamp;
  }

  public VehicleStatusMessage setProcStateTimestamp(Instant procStateTimestamp) {
    this.procStateTimestamp = requireNonNull(procStateTimestamp, "procStateTimestamp");
    return this;
  }

  public List<List<String>> getAllocatedResources() {
    return allocatedResources;
  }

  public VehicleStatusMessage setAllocatedResources(List<List<String>> allocatedResources) {
    this.allocatedResources = requireNonNull(allocatedResources, "allocatedResources");
    return this;
  }

  public List<List<String>> getClaimedResources() {
    return claimedResources;
  }

  public VehicleStatusMessage setClaimedResources(List<List<String>> claimedResources) {
    this.claimedResources = requireNonNull(claimedResources, "claimedResources");
    return this;
  }

  public List<AcceptableOrderTypeTO> getAcceptableOrderTypes() {
    return acceptableOrderTypes;
  }

  public VehicleStatusMessage setAcceptableOrderTypes(
      List<AcceptableOrderTypeTO> acceptableOrderTypes
  ) {
    this.acceptableOrderTypes = requireNonNull(acceptableOrderTypes, "acceptableOrderTypes");
    return this;
  }

  @Nonnull
  public String getEnvelopeKey() {
    return envelopeKey;
  }

  public VehicleStatusMessage setEnvelopeKey(
      @Nonnull
      String envelopeKey
  ) {
    this.envelopeKey = envelopeKey;
    return this;
  }

  /**
   * A precise position of a vehicle.
   */
  public static class PrecisePosition {

    private long x;

    private long y;

    private long z;

    /**
     * Creates a new instance.
     */
    public PrecisePosition() {
    }

    /**
     * Creates a new instance.
     *
     * @param x x value
     * @param y y value
     * @param z z value
     */
    public PrecisePosition(long x, long y, long z) {
      this.x = x;
      this.y = y;
      this.z = z;
    }

    public long getX() {
      return x;
    }

    public void setX(long x) {
      this.x = x;
    }

    public long getY() {
      return y;
    }

    public void setY(long y) {
      this.y = y;
    }

    public long getZ() {
      return z;
    }

    public void setZ(long z) {
      this.z = z;
    }
  }
}
