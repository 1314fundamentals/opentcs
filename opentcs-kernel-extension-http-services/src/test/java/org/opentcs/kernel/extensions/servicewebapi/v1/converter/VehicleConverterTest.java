// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.aMapWithSize;
import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.awt.Color;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.access.to.model.VehicleCreationTO;
import org.opentcs.data.model.AcceptableOrderType;
import org.opentcs.data.model.BoundingBox;
import org.opentcs.data.model.Point;
import org.opentcs.data.model.Pose;
import org.opentcs.data.model.Triple;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.model.Vehicle.EnergyLevelThresholdSet;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.GetVehicleResponseTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.getevents.VehicleStatusMessage;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel.VehicleTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.AcceptableOrderTypeTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.BoundingBoxTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.CoupleTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.PropertyTO;
import org.opentcs.util.Colors;

/**
 * Tests for {@link VehicleConverter}.
 */
class VehicleConverterTest {

  private VehicleConverter vehicleConverter;

  private Map<String, String> propertyMap;
  private List<PropertyTO> propertyTOList;
  private List<Property> propertyList;

  @BeforeEach
  void setUp() {
    PropertyConverter propertyConverter = mock();
    vehicleConverter = new VehicleConverter(propertyConverter, new AcceptableOrderTypeConverter());

    propertyMap = Map.of("some-key", "some-value");
    propertyTOList = List.of(new PropertyTO("some-key", "some-value"));
    propertyList = List.of(new Property("some-key", "some-value"));
    Set<AcceptableOrderType> acceptableOrderTypes = Set.of(new AcceptableOrderType("order-1", 0));
    List<AcceptableOrderTypeTO> acceptableOrderTypeList = List.of(
        new AcceptableOrderTypeTO("order-1", 0)
    );
    when(propertyConverter.toPropertyTOs(propertyMap)).thenReturn(propertyTOList);
    when(propertyConverter.toProperties(propertyMap)).thenReturn(propertyList);
    when(propertyConverter.toPropertyMap(propertyTOList)).thenReturn(propertyMap);
  }

  @Test
  void checkToVehicleCreationTOs() {
    VehicleTO vehicleTo = new VehicleTO("V1")
        .setBoundingBox(new BoundingBoxTO(500, 100, 700, new CoupleTO(0, 0)))
        .setEnergyLevelGood(90)
        .setEnergyLevelCritical(30)
        .setEnergyLevelFullyRecharged(90)
        .setEnergyLevelSufficientlyRecharged(30)
        .setMaxVelocity(1000)
        .setMaxReverseVelocity(1000)
        .setLayout(new VehicleTO.Layout())
        .setProperties(propertyTOList);

    List<VehicleCreationTO> result = vehicleConverter.toVehicleCreationTOs(List.of(vehicleTo));

    assertThat(result, hasSize(1));
    assertThat(result.get(0).getName(), is("V1"));
    assertThat(result.get(0).getBoundingBox().getLength(), is(500L));
    assertThat(result.get(0).getBoundingBox().getWidth(), is(100L));
    assertThat(result.get(0).getBoundingBox().getHeight(), is(700L));
    assertThat(result.get(0).getBoundingBox().getReferenceOffset().getX(), is(0L));
    assertThat(result.get(0).getBoundingBox().getReferenceOffset().getY(), is(0L));
    assertThat(result.get(0).getEnergyLevelThresholdSet().getEnergyLevelGood(), is(90));
    assertThat(result.get(0).getEnergyLevelThresholdSet().getEnergyLevelCritical(), is(30));
    assertThat(result.get(0).getEnergyLevelThresholdSet().getEnergyLevelFullyRecharged(), is(90));
    assertThat(
        result.get(0).getEnergyLevelThresholdSet().getEnergyLevelSufficientlyRecharged(), is(30)
    );
    assertThat(result.get(0).getMaxVelocity(), is(1000));
    assertThat(result.get(0).getMaxReverseVelocity(), is(1000));
    assertThat(result.get(0).getLayout().getRouteColor(), is(Colors.decodeFromHexRGB("#00FF00")));
    assertThat(result.get(0).getProperties(), is(aMapWithSize(1)));
    assertThat(result.get(0).getProperties(), is(propertyMap));
  }

  @Test
  void checkToVehicleTOs() {
    Vehicle vehicle = new Vehicle("V1")
        .withBoundingBox(new BoundingBox(500, 100, 700))
        .withEnergyLevelThresholdSet(new EnergyLevelThresholdSet(30, 90, 30, 90))
        .withMaxVelocity(1000)
        .withMaxReverseVelocity(1000)
        .withLayout(new Vehicle.Layout())
        .withProperties(propertyMap);

    List<VehicleTO> result = vehicleConverter.toVehicleTOs(Set.of(vehicle));

    assertThat(result, hasSize(1));
    assertThat(result.get(0).getName(), is("V1"));
    assertThat(result.get(0).getBoundingBox().getLength(), is(500L));
    assertThat(result.get(0).getBoundingBox().getWidth(), is(100L));
    assertThat(result.get(0).getBoundingBox().getHeight(), is(700L));
    assertThat(result.get(0).getBoundingBox().getReferenceOffset().getX(), is(0L));
    assertThat(result.get(0).getBoundingBox().getReferenceOffset().getY(), is(0L));
    assertThat(result.get(0).getEnergyLevelGood(), is(90));
    assertThat(result.get(0).getEnergyLevelCritical(), is(30));
    assertThat(result.get(0).getEnergyLevelFullyRecharged(), is(90));
    assertThat(result.get(0).getEnergyLevelSufficientlyRecharged(), is(30));
    assertThat(result.get(0).getMaxVelocity(), is(1000));
    assertThat(result.get(0).getMaxReverseVelocity(), is(1000));
    assertThat(result.get(0).getLayout().getRouteColor(), is(Colors.encodeToHexRGB(Color.RED)));
    assertThat(result.get(0).getProperties(), hasSize(1));
    assertThat(result.get(0).getProperties(), is(propertyTOList));
  }

  @Test
  void checkToGetVehicleResponseTO() {
    Vehicle vehicle = new Vehicle("V1")
        .withBoundingBox(new BoundingBox(500, 100, 700))
        .withEnergyLevelThresholdSet(new EnergyLevelThresholdSet(30, 90, 30, 90))
        .withEnergyLevel(50)
        .withIntegrationLevel(Vehicle.IntegrationLevel.TO_BE_UTILIZED)
        .withPaused(true)
        .withProcState(Vehicle.ProcState.PROCESSING_ORDER)
        .withTransportOrder(new TransportOrder("some-order", List.of()).getReference())
        .withCurrentPosition(new Point("some-point").getReference())
        .withPose(new Pose(new Triple(1, 2, 3), 4))
        .withState(Vehicle.State.EXECUTING)
        .withAllocatedResources(List.of(Set.of(new Point("some-other-point").getReference())))
        .withClaimedResources(List.of(Set.of(new Point("yet-another-point").getReference())))
        .withEnvelopeKey("some-key")
        .withAcceptableOrderTypes(Set.of(new AcceptableOrderType("some-type", 3)))
        .withMaxVelocity(1000)
        .withMaxReverseVelocity(1000)
        .withLayout(new Vehicle.Layout())
        .withProperties(propertyMap);

    GetVehicleResponseTO result = vehicleConverter.toGetVehicleResponseTO(vehicle);

    assertThat(result.getName(), is("V1"));
    assertThat(result.getBoundingBox().getLength(), is(500L));
    assertThat(result.getBoundingBox().getWidth(), is(100L));
    assertThat(result.getBoundingBox().getHeight(), is(700L));
    assertThat(result.getBoundingBox().getReferenceOffset().getX(), is(0L));
    assertThat(result.getBoundingBox().getReferenceOffset().getY(), is(0L));
    assertThat(result.getEnergyLevelGood(), is(90));
    assertThat(result.getEnergyLevelCritical(), is(30));
    assertThat(result.getEnergyLevelFullyRecharged(), is(90));
    assertThat(result.getEnergyLevelSufficientlyRecharged(), is(30));
    assertThat(result.getEnergyLevel(), is(50));
    assertThat(result.getIntegrationLevel(), is(Vehicle.IntegrationLevel.TO_BE_UTILIZED));
    assertThat(result.isPaused(), is(true));
    assertThat(result.getProcState(), is(Vehicle.ProcState.PROCESSING_ORDER));
    assertThat(result.getTransportOrder(), is("some-order"));
    assertThat(result.getCurrentPosition(), is("some-point"));
    assertThat(result.getPrecisePosition().getX(), is(1L));
    assertThat(result.getPrecisePosition().getY(), is(2L));
    assertThat(result.getPrecisePosition().getZ(), is(3L));
    assertThat(result.getOrientationAngle(), is(4.0));
    assertThat(result.getState(), is(Vehicle.State.EXECUTING));
    assertThat(result.getAllocatedResources(), is(List.of(List.of("some-other-point"))));
    assertThat(result.getClaimedResources(), is(List.of(List.of("yet-another-point"))));
    assertThat(result.getEnvelopeKey(), is("some-key"));
    assertThat(
        result.getAcceptableOrderTypes(),
        is(List.of(new AcceptableOrderTypeTO("some-type", 3)))
    );
    assertThat(result.getProperties(), is(aMapWithSize(1)));
    assertThat(result.getProperties(), is(propertyMap));
  }

  @Test
  void checkToVehicleStatusMessage() {
    Vehicle vehicle = new Vehicle("V1")
        .withBoundingBox(new BoundingBox(500, 100, 700))
        .withEnergyLevelThresholdSet(new EnergyLevelThresholdSet(30, 90, 30, 90))
        .withEnergyLevel(50)
        .withIntegrationLevel(Vehicle.IntegrationLevel.TO_BE_UTILIZED)
        .withPaused(true)
        .withProcState(Vehicle.ProcState.PROCESSING_ORDER)
        .withTransportOrder(new TransportOrder("some-order", List.of()).getReference())
        .withCurrentPosition(new Point("some-point").getReference())
        .withPose(new Pose(new Triple(1, 2, 3), 4))
        .withState(Vehicle.State.EXECUTING)
        .withAllocatedResources(List.of(Set.of(new Point("some-other-point").getReference())))
        .withClaimedResources(List.of(Set.of(new Point("yet-another-point").getReference())))
        .withEnvelopeKey("some-key")
        .withAcceptableOrderTypes(Set.of(new AcceptableOrderType("some-type", 3)))
        .withMaxVelocity(1000)
        .withMaxReverseVelocity(1000)
        .withLayout(new Vehicle.Layout())
        .withProperties(propertyMap);

    VehicleStatusMessage message = vehicleConverter.toVehicleStatusMessage(
        vehicle, 555, Instant.EPOCH
    );

    assertThat(message.getVehicleName(), is("V1"));
    assertThat(message.getBoundingBox().getLength(), is(500L));
    assertThat(message.getBoundingBox().getWidth(), is(100L));
    assertThat(message.getBoundingBox().getHeight(), is(700L));
    assertThat(message.getBoundingBox().getReferenceOffset().getX(), is(0L));
    assertThat(message.getBoundingBox().getReferenceOffset().getY(), is(0L));
    assertThat(message.getEnergyLevelGood(), is(90));
    assertThat(message.getEnergyLevelCritical(), is(30));
    assertThat(message.getEnergyLevelFullyRecharged(), is(90));
    assertThat(message.getEnergyLevelSufficientlyRecharged(), is(30));
    assertThat(message.getEnergyLevel(), is(50));
    assertThat(message.getIntegrationLevel(), is(Vehicle.IntegrationLevel.TO_BE_UTILIZED));
    assertThat(message.isPaused(), is(true));
    assertThat(message.getProcState(), is(Vehicle.ProcState.PROCESSING_ORDER));
    assertThat(message.getTransportOrderName(), is("some-order"));
    assertThat(message.getPrecisePosition().getX(), is(1L));
    assertThat(message.getPrecisePosition().getY(), is(2L));
    assertThat(message.getPrecisePosition().getZ(), is(3L));
    assertThat(message.getOrientationAngle(), is(4.0));
    assertThat(message.getState(), is(Vehicle.State.EXECUTING));
    assertThat(message.getAllocatedResources(), is(List.of(List.of("some-other-point"))));
    assertThat(message.getClaimedResources(), is(List.of(List.of("yet-another-point"))));
    assertThat(message.getEnvelopeKey(), is("some-key"));
    assertThat(
        message.getAcceptableOrderTypes(),
        is(List.of(new AcceptableOrderTypeTO("some-type", 3)))
    );
    assertThat(message.getProperties().size(), is(1));
    assertThat(message.getProperties().getFirst().getKey(), is("some-key"));
    assertThat(message.getProperties().getFirst().getValue(), is("some-value"));
  }
}
