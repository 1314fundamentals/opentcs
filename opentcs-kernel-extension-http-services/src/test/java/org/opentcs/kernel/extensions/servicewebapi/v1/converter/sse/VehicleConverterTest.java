// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter.sse;

import static org.assertj.core.api.Assertions.assertThat;

import java.awt.Color;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.assertj.core.groups.Tuple;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.opentcs.data.ObjectHistory;
import org.opentcs.data.model.AcceptableOrderType;
import org.opentcs.data.model.BoundingBox;
import org.opentcs.data.model.Couple;
import org.opentcs.data.model.Point;
import org.opentcs.data.model.Pose;
import org.opentcs.data.model.Triple;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.OrderSequence;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.drivers.vehicle.LoadHandlingDevice;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.TimestampScrubber;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.ObjectHistoryTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.VehicleEventTO;

/**
 * Tests for {@link VehicleConverter}.
 */
class VehicleConverterTest {

  private JsonBinder jsonBinder;
  private VehicleConverter converter;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
    converter = new VehicleConverter();
  }

  @Test
  void convert() {
    Vehicle vehicle = new Vehicle("vehicle-1")
        .withProperties(Map.of("key-1", "value-1", "key-2", "value-2"))
        .withHistoryEntry(
            new ObjectHistory.Entry(Instant.EPOCH, "event-code-1", List.of("supplement-1"))
        )
        .withBoundingBox(
            new BoundingBox(1, 2, 3)
                .withReferenceOffset(new Couple(4, 5))
        )
        .withEnergyLevelThresholdSet(new Vehicle.EnergyLevelThresholdSet(1, 2, 3, 4))
        .withMaxVelocity(200)
        .withMaxReverseVelocity(100)
        .withRechargeOperation("recharge-operation")
        .withProcState(Vehicle.ProcState.IDLE)
        .withTransportOrder(new TransportOrder("transport-order-1", List.of()).getReference())
        .withOrderSequence(new OrderSequence("order-sequence-1").getReference())
        .withAcceptableOrderTypes(
            Set.of(
                new AcceptableOrderType("acceptable-order-type-1", 2),
                new AcceptableOrderType("acceptable-order-type-2", 5)
            )
        )
        .withClaimedResources(
            List.of(
                Set.of(
                    new Point("point-1").getReference(),
                    new Point("point-2").getReference()
                ),
                Set.of(
                    new Point("point-3").getReference(),
                    new Point("point-4").getReference()
                )
            )
        )
        .withAllocatedResources(
            List.of(
                Set.of(
                    new Point("point-5").getReference(),
                    new Point("point-6").getReference()
                ),
                Set.of(
                    new Point("point-7").getReference(),
                    new Point("point-8").getReference()
                )
            )
        )
        .withState(Vehicle.State.IDLE)
        .withIntegrationLevel(Vehicle.IntegrationLevel.TO_BE_UTILIZED)
        .withPaused(true)
        .withCurrentPosition(new Point("point-5").getReference())
        .withPose(new Pose(new Triple(1, 2, 3), 4.5))
        .withEnergyLevel(42)
        .withLoadHandlingDevices(List.of(new LoadHandlingDevice("lhd-1", true)))
        .withEnvelopeKey("envelope-key-1")
        .withLayout(
            new Vehicle.Layout()
                .withRouteColor(new Color(20, 40, 60))
        );

    VehicleEventTO.VehicleTO result = converter.convert(vehicle);

    Approvals.verify(
        jsonBinder.toJson(result),
        // Scrub timestamps in the result's JSON representation as some of them may be generated
        // during test runtime, making it impossible to use ApprovalTests.
        new Options(new TimestampScrubber())
    );
  }

  /**
   * Ensures that timestamps are converted correctly. (Testing this is necessary, since timestamps
   * are scrubbed in the ApprovalTest and therefore their conversion is not covered there.)
   */
  @Test
  void convertsTimestampsCorrectly() {
    Instant time = Instant.parse("2025-07-25T12:03:00Z");
    Vehicle vehicle = new Vehicle("vehicle-1")
        .withHistoryEntry(
            new ObjectHistory.Entry(time, "event-code-1", List.of("supplement-1"))
        );

    VehicleEventTO.VehicleTO result = converter.convert(vehicle);

    assertThat(result.getHistory().getEntries())
        .extracting(
            ObjectHistoryTO.ObjectHistoryEntryTO::getTimestamp,
            ObjectHistoryTO.ObjectHistoryEntryTO::getEventCode,
            ObjectHistoryTO.ObjectHistoryEntryTO::getSupplements
        )
        .contains(Tuple.tuple(time, "event-code-1", List.of("supplement-1")));
    assertThat(result.getState().getTimestamp()).isEqualTo(vehicle.getStateTimestamp());
    assertThat(result.getProcState().getTimestamp()).isEqualTo(vehicle.getProcStateTimestamp());
  }

  @ParameterizedTest
  @EnumSource(Vehicle.State.class)
  void convertsAllStates(Vehicle.State state) {
    Vehicle vehicle = new Vehicle("vehicle-1").withState(state);

    VehicleEventTO.VehicleTO result = converter.convert(vehicle);

    VehicleEventTO.StateTO expectedState = switch (state) {
      case UNKNOWN -> VehicleEventTO.StateTO.UNKNOWN;
      case UNAVAILABLE -> VehicleEventTO.StateTO.UNAVAILABLE;
      case ERROR -> VehicleEventTO.StateTO.ERROR;
      case IDLE -> VehicleEventTO.StateTO.IDLE;
      case EXECUTING -> VehicleEventTO.StateTO.EXECUTING;
      case CHARGING -> VehicleEventTO.StateTO.CHARGING;
    };
    assertThat(result.getState().getState()).isEqualTo(expectedState);
  }

  @ParameterizedTest
  @EnumSource(Vehicle.ProcState.class)
  void convertsAllProcStates(Vehicle.ProcState procState) {
    Vehicle vehicle = new Vehicle("vehicle-1").withProcState(procState);

    VehicleEventTO.VehicleTO result = converter.convert(vehicle);

    VehicleEventTO.ProcStateTO expectedProcState = switch (procState) {
      case IDLE -> VehicleEventTO.ProcStateTO.IDLE;
      case AWAITING_ORDER -> VehicleEventTO.ProcStateTO.AWAITING_ORDER;
      case PROCESSING_ORDER -> VehicleEventTO.ProcStateTO.PROCESSING_ORDER;
    };
    assertThat(result.getProcState().getProcState()).isEqualTo(expectedProcState);
  }

  @ParameterizedTest
  @EnumSource(Vehicle.IntegrationLevel.class)
  void convertsAllIntegrationLevels(Vehicle.IntegrationLevel integrationLevel) {
    Vehicle vehicle = new Vehicle("vehicle-1").withIntegrationLevel(integrationLevel);

    VehicleEventTO.VehicleTO result = converter.convert(vehicle);

    VehicleEventTO.IntegrationLevelTO expectedIntegrationLevel = switch (integrationLevel) {
      case TO_BE_IGNORED -> VehicleEventTO.IntegrationLevelTO.TO_BE_IGNORED;
      case TO_BE_NOTICED -> VehicleEventTO.IntegrationLevelTO.TO_BE_NOTICED;
      case TO_BE_RESPECTED -> VehicleEventTO.IntegrationLevelTO.TO_BE_RESPECTED;
      case TO_BE_UTILIZED -> VehicleEventTO.IntegrationLevelTO.TO_BE_UTILIZED;
    };
    assertThat(result.getIntegrationLevel()).isEqualTo(expectedIntegrationLevel);
  }
}
