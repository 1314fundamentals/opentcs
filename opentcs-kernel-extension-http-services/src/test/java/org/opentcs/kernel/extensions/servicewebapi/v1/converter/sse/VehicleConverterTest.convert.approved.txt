{
  "name" : "vehicle-1",
  "properties" : {
    "key-1" : "value-1",
    "key-2" : "value-2"
  },
  "history" : {
    "entries" : [ {
      "timestamp" : "[Timestamp]",
      "eventCode" : "event-code-1",
      "supplements" : [ "supplement-1" ]
    } ]
  },
  "boundingBox" : {
    "length" : 1,
    "width" : 2,
    "height" : 3,
    "referenceOffset" : {
      "x" : 4,
      "y" : 5
    }
  },
  "energyLevelThresholdSet" : {
    "energyLevelCritical" : 1,
    "energyLevelGood" : 2,
    "energyLevelFullyRecharged" : 4,
    "energyLevelSufficientlyRecharged" : 3
  },
  "energyLevel" : 42,
  "maxVelocity" : 200,
  "maxReverseVelocity" : 100,
  "rechargeOperation" : "recharge-operation",
  "loadHandlingDevices" : [ {
    "label" : "lhd-1",
    "full" : true
  } ],
  "state" : {
    "state" : "IDLE",
    "timestamp" : "[Timestamp]"
  },
  "procState" : {
    "procState" : "IDLE",
    "timestamp" : "[Timestamp]"
  },
  "integrationLevel" : "TO_BE_UTILIZED",
  "paused" : true,
  "transportOrder" : "transport-order-1",
  "orderSequence" : "order-sequence-1",
  "acceptableOrderTypes" : [ {
    "name" : "acceptable-order-type-1",
    "priority" : 2
  }, {
    "name" : "acceptable-order-type-2",
    "priority" : 5
  } ],
  "claimedResources" : [ [ "point-1", "point-2" ], [ "point-3", "point-4" ] ],
  "allocatedResources" : [ [ "point-5", "point-6" ], [ "point-7", "point-8" ] ],
  "currentPosition" : "point-5",
  "pose" : {
    "position" : {
      "x" : 1,
      "y" : 2,
      "z" : 3
    },
    "orientationAngle" : 4.5
  },
  "envelopeKey" : "envelope-key-1",
  "layout" : {
    "routeColor" : {
      "red" : 20,
      "green" : 40,
      "blue" : 60
    }
  }
}